pkg/golinters/runtime_config.go

```diff
@@ -0,0 +1,263 @@
+		if cb, ok := t.X.(*ast.CallExpr); ok {
+			for _, arg := range cb.Args {
+				validateAssignConversion(arg, argName, defType, pass)
+			}
+		}
+	}
+}
+
+func validateAssignConversion(arg ast.Expr, argName, defType string, pass *analysis.Pass) {
+	if expr, ok := arg.(*ast.TypeAssertExpr); ok {
+		if ident, ok := expr.X.(*ast.Ident); ok && ident.Name == argName {
+			if se, ok := expr.Type.(*ast.StarExpr); !ok {
+				pass.Reportf(expr.Type.Pos(), "the configuration object (arg %s) is a reference, add '*' to fix type conversion", argName)
+			} else if typeConv, ok := se.X.(*ast.Ident); ok {
+				checkSameType(typeConv, argName, defType, pass)
+			}
+		}
+	}
+}
+
+func checkSameType(ident *ast.Ident, argName, defType string, pass *analysis.Pass) {
+	if !strings.HasSuffix(defType, ident.Obj.Name) {
+		pass.Reportf(ident.Pos(),
+			"the configuration object (arg %s) has to match the type of the default argument, expected type: %s but found %s",
+			argName, defType, ident.Obj.Name,
+		)
+	}
+}
+
+func validateFieldSerialization(pass *analysis.Pass, pos token.Pos, config *types.Struct, configTypeName string) {
+	for i := 0; i < config.NumFields(); i++ {
+		field := config.Field(i)
+		tag := config.Tag(i)
+		// oneOf's don't specify a json tag but the serializer 'correctly' uses the camelCase name
+		isOneOf := strings.Contains(tag, "protobuf_oneof")
+
+		// ignore missing tags for unexported fields and oneOfs
+		if !field.Exported() || isOneOf {
+			continue
+		}
+		ft := getStructType(field.Type())
+		// recurse into complex types
+		if field.Embedded() {
+			validateFieldSerialization(pass, pos, ft, fmt.Sprintf("%s embedded in %s", field.Type(), configTypeName))
+			continue
+		}
+		hasJSONTag := strings.Contains(tag, "json")
+
+		if !hasJSONTag || !hasValidJSONTag(tag) {
+			pass.Reportf(pos,
+				`runtimeConfigurations must specify a json field name in camelCase format (e.g. json:"fieldName") for 
+exported fields, found field '%s' of '%s' using '%s'`,
+				field.Name(), configTypeName, tag,
+			)
+		}
+		if ft != nil {
+			validateFieldSerialization(pass, pos, ft, fmt.Sprintf("%s in %s", field.Type(), configTypeName))
+		}
+	}
+}
+
+func getStructType(t types.Type) *types.Struct {
+	switch t := t.(type) {
+	case *types.Pointer:
+		return getStructType(t.Elem())
+	case *types.Named:
+		return getStructType(t.Underlying())
+	case *types.Struct:
+		return t
+	default:
+		return nil
+	}
+}
+
+func hasValidJSONTag(tag string) bool {
+	// protos contain proto and reg. json tag, proto is preferred
+	matches := jsonProtoReg.FindStringSubmatch(tag)
+	expectedMatches := 2
+	if len(matches) != expectedMatches {
+		matches = jsonReg.FindStringSubmatch(tag)
+		if len(matches) != expectedMatches {
+			return false
+		}
+	}
+	for _, m := range strings.Split(matches[1], ",") {
+		if m == "" || m == "omitempty" || m == "-" || m == "string" {
+			continue
+		}
+		return camelCase(m) == m
+	}
+	return false
+}
+
+func camelCase(fieldName string) string {
```

Comment:
Could maybe use this package? https://pkg.go.dev/github.com/fatih/camelcase

e.g. `return strings.join(camelcase.Split(tag), "") == tag`

Reply:
Not a fan of external dependencies for 'simple' utils but can do if you feel strongly about it.